### PR TITLE
cmd/snap-bootstrap: measure epoch and model before unlocking encrypted data

### PR DESCRIPTION
**Pull request by bboozzoo**
Measure the epoch and system model before attempting to unlock ubuntu-data partition.

### CODE REVIEW

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 1]
```diff
@@ -325,9 +365,13 @@ func generateMountsModeRun() error {
 		return err
 	}
 	if !isDataMounted {
-		name := filepath.Base(boot.InitramfsUbuntuDataDir)
+		if isDeviceEncrypted("ubuntu-data") {
+			if err := measureRunModeSystem(); err != nil {
```
Comment (chrisccoulson) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:369]: The measurements need to happen on all boots if a TPM is detected, regardless of whether the system being booted has an encrypted volume.

Comment (pedronis) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:369]: @chrisccoulson should the same apply to the locking of keys ?

Comment (chrisccoulson) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:369]: Yes, the same applies to locking of keys

**Comment by cmatsuoka** [comment]
Thanks Maciek. Some refactoring still needed (to measure whenever TPM is available), but the measurements are correct and unsealing work as expected when tested with an encryption key sealed by PR #8531.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 2]
```diff
@@ -245,6 +252,22 @@ func generateMountsCommonInstallRecover(recoverySystem string) (allMounted bool,
 		return false, nil
 	}
 
+	// 1a. measure model
+	err = stampedAction("recovery-model-measure", func() error {
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:256]: maybe we want to make the stamp  recoverySystem+"-model-measured"

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 3]
```diff
@@ -304,6 +327,55 @@ func generateMountsCommonInstallRecover(recoverySystem string) (allMounted bool,
 	return true, nil
 }
 
+func loadModelFrom(where string) (*asserts.Model, error) {
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:330]: s/load/read/

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 4]
```diff
@@ -319,15 +391,29 @@ func generateMountsModeRun() error {
 		}
 	}
 
+	// 1.1a measure model
+	err := stampedAction("boot-model-measure", func() error {
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:395]: run-model-measured

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 5]
```diff
@@ -507,11 +593,36 @@ func generateMountsModeRun() error {
...
+	}
+	if err := action(); err != nil {
+		return err
+	}
+	return ioutil.WriteFile(stampFile, nil, 0644)
+}
+
 func generateInitramfsMounts() error {
+	// Ensure there is a very early initial measurement
+	err := stampedAction("initial-measure", func() error {
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:614]: secboot-epoch-measured

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 6]
```diff
@@ -528,13 +639,17 @@ var (
 	secbootLockAccessToSealedKeys = secboot.LockAccessToSealedKeys
 )
 
+func isDeviceEncrypted(name string) bool {
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:642]: maybe this can return  (ok bool, encdev string) ?

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 7]
```diff
@@ -319,15 +391,29 @@ func generateMountsModeRun() error {
...
+			if err != nil {
+				return fmt.Errorf("cannot load run mode model: %v", err)
+			}
+			return secMeasureModel(tpm, model)
+		})
+	})
+	if err != nil {
+		return err
+	}
+
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:408]: we should add a TODO:UC20 abot x-checking the model after data is mounted

Comment (bboozzoo) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:408]: added

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 8]
```diff
@@ -245,6 +250,23 @@ func generateMountsCommonInstallRecover(recoverySystem string) (allMounted bool,
 		return false, nil
 	}
 
+	// 1a. measure model
+	err = stampedAction(fmt.Sprintf("%s-model-measure", recoverySystem), func() error {
+		return measureWhenPossible(func(tpm *secboot.TPMConnection) error {
+			// TODO:UC20: consider doing the measurement later and
+			// getting the model from opened system seed
```
Comment (mvo5) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:257]: I'm not sure I understand this comment, this is reading from "ubuntu-seed", what would be the difference?

Comment (bboozzoo) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:257]: A chance for potential refactor. Couple lines down we have `recoverySystemEssentialSnaps()` which opens the system seed. The seed itself has some convenience methods for accessing the model.

Comment (pedronis) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:257]: the model at this point is not verified

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 9]
```diff
@@ -245,6 +250,23 @@ func generateMountsCommonInstallRecover(recoverySystem string) (allMounted bool,
 		return false, nil
 	}
 
+	// 1a. measure model
+	err = stampedAction(fmt.Sprintf("%s-model-measure", recoverySystem), func() error {
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:254]: this should be "*-measured"  but seems there's no test that would double check for this ATM

**Comment by cmatsuoka** [comment]
On qemu/swtpm the current implementation blocks for some time (up to 2 minutes on my machine if left untouched) in the first attempt to connect to TPM because we don't have enough entropy. This started to happen after the TPM connection for measurements was moved to an earlier point in execution. @chrisccoulson tells me that we shouldn't hit issues like this on real hardware because the kernel is configured to trust the CPU's hrng to populate its own entropy pool so there should always be entropy, even during early boot. Having it working on a VM would be useful however for demo/tests.

File: cmd/snap-bootstrap/initramfs_mounts_state.go [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+// initramfsMountsState helps tracking the state and progress
+// of the mounting driving process.
+type initramfsMountsState interface {
+	// Model returns the verified model from the seed (only for
+	// modes other than run).
+	Model() (*asserts.Model, error)
+
+	// RecoverySystemEssentialSnaps returns the verified essential
+	// snaps from the recoverySystem, if that is "" the implied one
+	// will be used (only for modes other than run).
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/initramfs_mounts_state.go:44]: this is slightly hard to parse, what about

```suggestion
	// RecoverySystemEssentialSnaps returns the verified essential
	// snaps from the recoverySystem. If the recoverySystem is "" the
	// implied one will be used (only for modes other than run).
```

File: cmd/snap-bootstrap/initramfs_mounts_state.go [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+	RecoverySystemEssentialSnaps(recoverySystem string, essentialTypes []snap.Type) ([]*seed.Snap, error)
+
+	// UnverifiedBootModel returns the unverified model from the boot
+	// partition for run mode.
+	UnverifiedBootModel() (*asserts.Model, error)
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/initramfs_mounts_state.go:49]: would the unverified model ever come from anywhere other than the boot partition? If not, it seems like this could be `UnverifiedModel()` instead

File: cmd/snap-bootstrap/initramfs_mounts_state.go [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+type initramfsMountsStateImpl struct {
+	mode           string
+	recoverySystem string
+
+	seed seed.EssentialMetaLoaderSeed
+}
+
+var errRunModeNoImpliedRecoverySystem = errors.New("internal error: no implied recovery system in run mode")
+
+// loadSeed open the seed and reads assertions once, setting mst.seed
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/initramfs_mounts_state.go:68]:
```suggestion
// loadSeed opens the seed and reads them into mst.seed; it does not re-read seed when called multiple times
```

File: cmd/snap-bootstrap/initramfs_mounts_state.go [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+	}
+
+	// load assertions into a temporary database
+	if err := seed20.LoadAssertions(nil, nil); err != nil {
+		return err
+	}
+
+	mst.seed = seed20
+	return nil
+
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/initramfs_mounts_state.go:98]: unneeded whitespace

```suggestion
```

File: cmd/snap-bootstrap/initramfs_mounts_state.go (hunk 11 above)
Comment (pedronis) [reply cmd/snap-bootstrap/initramfs_mounts_state.go:49]: I really hope that's the only situation where we handle an unverified model, so the method name should stay. Maybe I need to clarify this somewhere.

Comment (anonymouse64) [reply cmd/snap-bootstrap/initramfs_mounts_state.go:49]: okay, that's fine
